dpctl/_sycl_context.pyx

```diff
@@ -24,28 +24,152 @@ from __future__ import print_function
 import logging
 from ._backend cimport (
     DPCTLSyclContextRef,
+    DPCTLSyclDeviceRef,
+    DPCTLContext_Create,
+    DPCTLContext_CreateFromDevices,
+    DPCTLContext_DeviceCount,
+    DPCTLContext_GetDevices,
+    DPCTLContext_Copy,
     DPCTLContext_Delete,
     DPCTLContext_AreEq,
+    DPCTLDevice_Delete,
+    DPCTLDevice_Copy,
+    DPCTLDeviceVectorRef,
+    DPCTLDeviceVector_CreateFromArray,
+    DPCTLDeviceVector_GetAt,
+    DPCTLDeviceVector_Size,
+    DPCTLDeviceVector_Delete,
+    error_handler_callback,
+    DPCTL_DeviceAndContextPair,
+    DPCTLDeviceMgr_GetDeviceAndContextPair,
 )
+from ._sycl_queue cimport default_async_error_handler
+from ._sycl_device cimport SyclDevice
+from cpython.mem cimport PyMem_Malloc, PyMem_Free
 
 __all__ = [
     "SyclContext",
 ]
 
 _logger = logging.getLogger(__name__)
 
+cdef class _SyclContext:
+    """ Data owner for SyclContext
+    """
+
+    def __dealloc__(self):
+        DPCTLContext_Delete(self._ctxt_ref)
 
-cdef class SyclContext:
+
+cdef class SyclContext(_SyclContext):
     """ Python wrapper class for cl::sycl::context.
     """
+    
+    @staticmethod
+    cdef void _init_helper(_SyclContext context, DPCTLSyclContextRef CRef):
+         context._ctxt_ref = CRef
+    
     @staticmethod
     cdef SyclContext _create (DPCTLSyclContextRef ctxt):
-        cdef SyclContext ret = SyclContext.__new__(SyclContext)
-        ret._ctxt_ref = ctxt
-        return ret
+        cdef _SyclContext ret = <_SyclContext>_SyclContext.__new__(_SyclContext)
+        SyclContext._init_helper(ret, ctxt)
+        return SyclContext(ret)
 
-    def __dealloc__ (self):
-        DPCTLContext_Delete(self._ctxt_ref)
+    cdef int _init_from__SyclContext(self, _SyclContext other):
+        self._ctxt_ref = DPCTLContext_Copy(other._ctxt_ref)
+        if (self._ctxt_ref is NULL):
+            return -1
+        return 0
+
+    cdef int _init_from_one_device(self, SyclDevice device, int props):
+        cdef DPCTLSyclDeviceRef DRef = device.get_device_ref()
+        cdef DPCTLSyclContextRef CRef = NULL
+        cdef error_handler_callback * eh_callback = \
+            <error_handler_callback *>&default_async_error_handler
+        cdef DPCTL_DeviceAndContextPair dev_ctx
+        # look up cached contexts for root devices first
+        dev_ctx = DPCTLDeviceMgr_GetDeviceAndContextPair(DRef)
+        if (dev_ctx.CRef is NULL) or (dev_ctx.DRef is NULL):
+            # look-up failed, create a new one
+            CRef = DPCTLContext_Create(DRef, eh_callback, props)
+            if (CRef is NULL):
+                return -1
+        else:
+            CRef = dev_ctx.CRef
```

Comment:
An issue that might occur is that the device (`DRef`) in the `CRef` found by the lookup will be different from the device ref in the `SyclDevice` argument. Will that cause an issue?

Reply:
We do not store the `SyclDevice` in the `SyclContext`. It will be possible to retrieve the device stored in the context.

When DPC++ properly supports device equality we wont need DeviceAndContextPair and the look-up could just return the context found.